drivers/xen/regions.c

```diff
@@ -0,0 +1,164 @@
+		     (uint8_t *)DT_REG_ADDR_BY_IDX(XEN_HYP_NODE, idx));
+
+DT_FOREACH_REG(XEN_HYP_NODE, EXT_BLOCK);
+
+#define EXT_REG_POINTER(node_id, idx)					\
+	{								\
+		.block = &region_##idx,					\
+		.phys_addr = DT_REG_ADDR_BY_IDX(XEN_HYP_NODE, idx),	\
+		.size = DT_REG_SIZE_BY_IDX(XEN_HYP_NODE, idx),		\
+	},
+
+/*
+ * TODO: For now we allocate buffer structures for DT_NUM_REGS, when we
+ * need a range 1...DT_FOREACH_REG. Although region 0 is ignored, it still
+ * takes stack space. This requires optimization
+ */
+static struct extended_regions ext_blocks[DT_NUM_REGS(XEN_HYP_NODE)] = {
+	DT_FOREACH_REG(XEN_HYP_NODE, EXT_REG_POINTER)
+};
+
+void *xen_region_get_pages(size_t nr_pages)
+{
+	int ret, i;
+	void *ptr;
+
+	/*
+	 * Locking ranges iteration, although each call sys_mem_blocks_alloc_contiguous
+	 * has it's own syncronisation, to avoid the situation of 2 parralel calls of
+	 * get_page and put_page on full buffers when put_page clears space in block 1
+	 * when get_page has already checked block 1. This will lead us to -ENOMEM while
+	 * block 1 space was freed already
+	 */
+	k_mutex_lock(&ext_regions_lock, K_FOREVER);
+
+	for (i = EXTENDED_REGIONS_IDX; i < DT_NUM_REGS(XEN_HYP_NODE); i++) {
+		ret = sys_mem_blocks_alloc_contiguous(ext_blocks[i].block, nr_pages, &ptr);
+		if (ret == 0) {
+			break;
+		} else if (ret && ret != -ENOMEM) {
+			LOG_ERR("Unable to allocate pages, ret = %d\n", ret);
+			ptr = NULL;
+		}
+	}
+
+	k_mutex_unlock(&ext_regions_lock);
+
+	return ptr;
+}
+
+void xen_region_put_pages(void *ptr, size_t nr_pages)
+{
+	int ret, i;
+	sys_mem_blocks_t *block = NULL;
+
+	k_mutex_lock(&ext_regions_lock, K_FOREVER);
+
+	for (i = EXTENDED_REGIONS_IDX; i < DT_NUM_REGS(XEN_HYP_NODE); i++) {
+		/* We can compare ptr to phys_addr because mapping is 1:1 */
+		if (((uint64_t)ptr >= ext_blocks[i].phys_addr &&
+		     (uint64_t)ptr < ext_blocks[i].phys_addr + ext_blocks[i].size) {
+			block = ext_blocks[i].block;
+		}
+	}
+
+	if (!block) {
+		LOG_ERR("Unable to find buffer, matching to addr %p\n", ptr);
+		goto unlock;
+	}
+
+	ret = sys_mem_blocks_free_contiguous(block, ptr, nr_pages);
+	if (ret) {
+		LOG_ERR("Unable to free pages, ret = %d\n", ret);
+		goto unlock;
+	}
+
+unlock:
+	k_mutex_unlock(&ext_regions_lock);
+}
+
+static int xen_regions_init(const struct device *d)
+{
+	int i;
+
+	ARG_UNUSED(d);
+
+	LOG_DBG("Initializing Xen Extended regions driver");
+
+	k_mutex_lock(&ext_regions_lock, K_FOREVER);
+
+	for (i = EXTENDED_REGIONS_IDX; i < DT_NUM_REGS(XEN_HYP_NODE); i++) {
+		/*
+		 * Using arch_mem_map to map the extended regions as 1:1.
+		 * The main idea of this functionality is to avoid using heap
+		 * and use 'unused memory' provided by XEN as extended_regions.
+		 * If using device_map call to map the space - then there is no
+		 * difference comparing to heap - we will get vaddr.
+		 * The idea is to allocate memory for the foreign mapping without
+		 * affecting Zephyr virtual address space.
```

Comment:
You are affecting virtual zephyr address space anyways, just by calling `arch_mem_map`